src/sentry/similarity.py

```diff
@@ -0,0 +1,245 @@
+                            desc=True,
+                            withscores=True,
+                        ),
+                        range(len(self.bands)),
+                    ) for key in keys
+                }
+
+            result = {}
+            for key, promises in responses.items():
+                # Resolve each promise, and scale the number of observations
+                # for each bucket to [0,1] value (the proportion of items
+                # observed in that band that belong to the bucket for the key.)
+                result[key] = map(
+                    lambda promise: scale_to_total(
+                        dict(promise.value)
+                    ),
+                    promises,
+                )
+
+            return result
+
+        def fetch_candidates(signature):
+            """Fetch all the similar candidates for a given signature."""
+            with self.cluster.map() as client:
+                responses = map(
+                    lambda (band, buckets): map(
+                        lambda bucket: client.smembers(
+                            b'{}:{}:{}:{}:{}'.format(self.namespace, scope, self.BUCKET_MEMBERSHIP, band, bucket)
+                        ),
+                        buckets,
+                    ),
+                    enumerate(signature),
+                )
+
+            # Resolve all of the promises for each band and reduce them into a
+            # single set per band.
+            return map(
+                lambda band: reduce(
+                    lambda values, promise: values | promise.value,
+                    band,
+                    set(),
+                ),
+                responses,
+            )
+
+        target_frequencies = fetch_bucket_frequencies([key])[key]
+
+        # Flatten the results of each band into a single set. (In the future we
+        # might want to change this to only calculate the similarity for keys
+        # that show up in some threshold number of bands.)
+        candidates = reduce(
+            lambda total, band: total | band,
+            fetch_candidates(target_frequencies),
+            set(),
+        )
+
+        return sorted(
+            map(
+                lambda (key, candidate_frequencies): (
+                    key,
+                    self.get_similarity(
+                        target_frequencies,
+                        candidate_frequencies,
+                    ),
+                ),
+                fetch_bucket_frequencies(candidates).items(),
+            ),
+            key=lambda (key, similarity): (similarity * -1, key),
+        )
+
+    def record(self, scope, key, characteristics):
+        """Records the presence of a set of characteristics within a group."""
+        with self.cluster.map() as client:
+            for band, buckets in enumerate(self.get_signature(characteristics)):
+                buckets = self.__format_buckets(buckets)
+                client.sadd(
```

Comment:
Yeah, my off-the-cuff thoughts is to wait and see if this is an issue before adding a cache layer on top of it (and evaluating whether or not it would be worth it.)